File: cplus/leetcode/202504/algo1770.cpp
#include <bits/stdc++.h>
using namespace std;

// 动态规划 -  区间 DP
class Solution {
public:
    int maximumScore(vector<int>& nums, vector<int>& multipliers) {
        // 自行解答，记忆化搜索。dp 还不知道怎么写
        int m = multipliers.size(), n = nums.size();

        vector<vector<int>> memo(m, vector<int>(m, -1));
        auto dfs = [&](this auto&& dfs, int i, int j) -> int{
            if (i+j==m) return 0;
            auto& res = memo[i][j];
            if (res != -1) return res;
            
            auto& mul = multipliers[i+j];
            return res = max(nums[i]*mul + dfs(i+1, j), nums[n-j-1]*mul + dfs(i, j+1));
        };
        return dfs(0,0);
    }
};

File: python/algo/202504/algo1770.py
from functools import cache
from typing import List

# 动态规划 - 8.2 其他区间 DP
class Solution:
    def maximumScore(self, nums: List[int], multipliers: List[int]) -> int:
        # 自行解答 n = len(nums)  n = 10^5
        m = len(multipliers)

        @cache
        def dfs(i:int, j:int) -> int:   # 关键是参数的选择 m 10^3
            # 避免用 n 相关参数从 0 开始从头，尾遍历 nums
            if i+j == m: return 0   # i + j == m

            mul = multipliers[i+j]
            return max(nums[i] * mul +dfs(i+1, j), nums[-j-1] * mul + dfs(i, j+1))

        return dfs(0, 0) 